src/hotspot/share/opto/loopnode.cpp

```diff
@@ -3631,6 +3631,17 @@ void IdealLoopTree::check_safepts(VectorSet &visited, Node_List &stack) {
             }
             // Skip to head of inner loop
             assert(_phase->is_dominator(_head, nlpt->_head), "inner head dominated by outer head");
+            if (_head == nlpt->_head) {
+              // this and nlpt (inner loop) have the same loop head. This should not happen because
+              // during beautify_loops we call merge_many_backedges. However, infinite loops may not
+              // have been attached to the loop-tree during build_loop_tree before beautify_loops,
+              // but then attached in the build_loop_tree afterwards, and so still have unmerged
+              // backedges. Check if we are indeed in an infinite subgraph, and terminate the scan,
+              // since we have reached the loop head of this.
+              assert(_head->as_Region()->is_in_infinite_subgraph(),
+                     "only expect unmerged backedges in infinite loops");
+              break;
+            }
             n = nlpt->_head;
```

Comment:
Just a small detail, I would move this above the new `if` such that you can use `n` in the check.